File: 102-BinaryTreeLevelOrderTraversal/102-BinaryTreeLevelOrderTraversal.py
# Last updated: 6/29/2026, 10:41:20 PM
1# Definition for a binary tree node.
2# class TreeNode:
3#     def __init__(self, val=0, left=None, right=None):
4#         self.val = val
5#         self.left = left
6#         self.right = right
7class Solution:
8    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
9        # BFS
10        if not root: return []
11        res = [[root.val]]
12        queue = deque()
13        queue.append(root)
14        while queue:
15            curr_len = len(queue)
16            level = []
17            for i in range(curr_len):
18                curr = queue.popleft()
19                if curr.left:
20                    level.append(curr.left.val)
21                    queue.append(curr.left)
22                if curr.right:
23                    level.append(curr.right.val)
24                    queue.append(curr.right)
25            if len(level) != 0:
26                res.append(level)
27        
28        return res
29
30        

File: 102-BinaryTreeLevelOrderTraversal/102-BinaryTreeLevelOrderTraversal.java
// Last updated: 8/16/2026, 9:52:34 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        // BFS
        List<List<Integer>> res = new ArrayList<>();
        if(root == null) return res;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            List<Integer> temp = new ArrayList<>();
            int qSize = queue.size();
            for(int i = 0; i < qSize; i ++) {
                TreeNode curr = queue.poll();
                temp.add(curr.val);

                if(curr.left != null) queue.offer(curr.left);
                if(curr.right != null) queue.offer(curr.right);
            }
            res.add(temp);
        }
        return res;
    }
}